internal/controller/namespacelabel_controller.go

```diff
@@ -0,0 +1,242 @@
+	r.Log.Info("Processing labels for Namespacelabel", "namespace", namespaceLabel.Namespace)
+
+	updatedLabels = make(map[string]string)
+	skippedLabels = make(map[string]string)
+	duplicateLabels = make(map[string]string)
+
+	if namespace.Labels == nil {
+		namespace.Labels = make(map[string]string)
+	}
+
+	for key, value := range namespaceLabel.Spec.Labels {
+		switch {
+		case protectedLabels[key] != "":
+			r.Log.Info("Skipping protected label", "key", key, "value", value)
+			skippedLabels[key] = value
+			r.Recorder.Event(namespaceLabel, corev1.EventTypeWarning, "ProtectedLabelSkipped", fmt.Sprintf("Label %s=%s is protected and was not applied", key, value))
+
+		case namespace.Labels[key] != "":
+			r.Log.Info("Skipping duplicate label", "key", key, "value", value)
+			duplicateLabels[key] = value
+			r.Recorder.Event(namespaceLabel, corev1.EventTypeWarning, "DuplicateLabelSkipped", fmt.Sprintf("Label %s=%s already exists with value %s", key, value, namespace.Labels[key]))
+
+		default:
+			r.Log.Info("Adding label", "key", key, "value", value)
+			updatedLabels[key] = value
+		}
+	}
+	return updatedLabels, skippedLabels, duplicateLabels
+}
+
+func (r *NamespacelabelReconciler) updateStatus(ctx context.Context, namespaceLabel *labelsv1.Namespacelabel, updatedLabels, skippedLabels, duplicateLabels map[string]string) error {
+	namespaceLabel.Status.AppliedLabels = updatedLabels
+	namespaceLabel.Status.SkippedLabels = skippedLabels
+
+	if len(skippedLabels) > 0 {
+		r.SetCondition(namespaceLabel, "LabelsSkipped", metav1.ConditionTrue, "ProtectedLabelsHandled", "Some labels were skipped because they are protected.")
+	} else {
+		r.SetCondition(namespaceLabel, "LabelsSkipped", metav1.ConditionFalse, "ProtectedLabelsHandled", "All labels were applied successfully; no protected labels were skipped.")
+	}
+
+	if len(duplicateLabels) > 0 {
+		r.SetCondition(namespaceLabel, "DuplicateLabels", metav1.ConditionTrue, "DuplicateLabelsHandled", "Some labels were not applied because they are duplicates.")
+	} else {
+		r.SetCondition(namespaceLabel, "DuplicateLabels", metav1.ConditionFalse, "DuplicateLabelsHandled", "All labels were unique and applied successfully.")
+	}
+
+	r.SetCondition(namespaceLabel, "LabelsApplied", metav1.ConditionTrue, "LabelsReconciled", "Labels reconciled successfully.")
+
+	if err := r.Status().Update(ctx, namespaceLabel); err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			r.Log.Info("Resource already deleted, skipping status update", "namespaceLabel", namespaceLabel.Name)
+			return nil
+		}
+		return fmt.Errorf("failed to update Namespacelabel status: %w", err)
+	}
+
+	return nil
+}
+
+func (r *NamespacelabelReconciler) ensureFinalizer(ctx context.Context, namespaceLabel *labelsv1.Namespacelabel) error {
+	if err := finalizer.Ensure(ctx, r.Client, namespaceLabel, r.Log); err != nil {
+		return err
+	}
+	return nil
```

Comment:
Can just rewrite this as:

```go
func (r *NamespacelabelReconciler) ensureFinalizer(ctx context.Context, namespaceLabel *labelsv1.Namespacelabel) error {
	return finalizer.Ensure(ctx, r.Client, namespaceLabel, r.Log)
}
```

You can skip this function and just use it directly from the caller